File: frontend/src/lib/user-preferences.ts
import { api } from "@/lib/api";

export interface UserPreferences {
  onboarding_completed: boolean;
  role: string | null;
  experience_level: string | null;
  primary_languages: string[];
  use_cases: string[];
  response_style: string | null;
  custom_context: string | null;
  memory_enabled: boolean;
  memory_compaction_day: number;
  memory_compaction_hour: number;
  memory_last_compacted_at: string | null;
  memory_next_run_at: string | null;
}

export interface UpdateUserPreferencesPayload {
  onboarding_completed?: boolean;
  role?: string | null;
  experience_level?: string | null;
  primary_languages?: string[];
  use_cases?: string[];
  response_style?: string | null;
  custom_context?: string | null;
  memory_enabled?: boolean;
  memory_compaction_day?: number;
  memory_compaction_hour?: number;
}

export async function getPreferences(): Promise<UserPreferences> {
  return api<UserPreferences>("/api/v1/user-preferences");
}

export async function updatePreferences(
  payload: UpdateUserPreferencesPayload
): Promise<UserPreferences> {
  return api<UserPreferences>("/api/v1/user-preferences", {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

File: backend/app/domain/entities/user_preferences.py
from dataclasses import dataclass, field
from datetime import datetime
from uuid import UUID


@dataclass
class UserPreferences:
    user_id: UUID
    onboarding_completed: bool = False
    role: str | None = None
    experience_level: str | None = None
    primary_languages: list[str] = field(default_factory=list)
    use_cases: list[str] = field(default_factory=list)
    response_style: str | None = None
    custom_context: str | None = None
    memory_enabled: bool = True
    memory_compaction_day: int = 0
    memory_compaction_hour: int = 3
    memory_last_compacted_at: datetime | None = None
    memory_next_run_at: datetime | None = None
